src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexReplacement.cs

```diff
@@ -113,82 +112,70 @@ public RegexReplacement(string rep, RegexNode concat, Hashtable _caps)
                 else
                 {
                     switch (-Specials - 1 - r)
                     { // special insertion patterns
                         case LeftPortion:
-                            sb.Append(match.GetLeftSubstring());
+                            vsb.Append(match.GetLeftSubstring());
                             break;
                         case RightPortion:
-                            sb.Append(match.GetRightSubstring());
+                            vsb.Append(match.GetRightSubstring());
                             break;
                         case LastGroup:
-                            sb.Append(match.LastGroupToStringImpl());
+                            vsb.Append(match.LastGroupToStringImpl());
                             break;
                         case WholeString:
-                            sb.Append(match.Text);
+                            vsb.Append(match.Text);
                             break;
                     }
                 }
             }
         }
 
-        /// <summary>
-        /// Given a Match, emits into the List<string> the evaluated
+       /// <summary>
+        /// Given a Match, emits into the ValueStringBuilder the evaluated
         /// Right-to-Left substitution pattern.
         /// </summary>
-        private void ReplacementImplRTL(List<string> al, Match match)
+        public void ReplacementImplRTL(ref ValueStringBuilder vsb, Match match)
         {
             for (int i = _rules.Count - 1; i >= 0; i--)
             {
                 int r = _rules[i];
                 if (r >= 0)  // string lookup
-                    al.Add(_strings[r]);
+                    vsb.AppendReversed(_strings[r]);
                 else if (r < -Specials) // group lookup
-                    al.Add(match.GroupToStringImpl(-Specials - 1 - r).ToString());
+                    vsb.AppendReversed(match.GroupToStringImpl(-Specials - 1 - r));
                 else
                 {
                     switch (-Specials - 1 - r)
                     { // special insertion patterns
                         case LeftPortion:
-                            al.Add(match.GetLeftSubstring().ToString());
+                            vsb.AppendReversed(match.GetLeftSubstring());
                             break;
                         case RightPortion:
-                            al.Add(match.GetRightSubstring().ToString());
+                            vsb.AppendReversed(match.GetRightSubstring());
                             break;
                         case LastGroup:
-                            al.Add(match.LastGroupToStringImpl().ToString());
+                            vsb.AppendReversed(match.LastGroupToStringImpl());
                             break;
                         case WholeString:
-                            al.Add(match.Text);
+                            vsb.AppendReversed(match.Text);
```

Comment:
Not for this specific location, but it might be worth from a perf perspective (in particular for these places where vsb is used in a loop), storing the vsb into a non-ref local (rather than a ref param) at the beginning of the method and then assigning it back to the ref param at the end, and see if that makes any difference.

Reply:
Tried it out and didn't see any noticeable improvements outside of the standard deviation.